site/src/job_queue.rs

```diff
@@ -33,10 +34,97 @@ async fn enqueue_master_commits(ctxt: &Arc<SiteCtxt>) {
     }
 }
 
+// This function is split for testing purposes as mocking out `SiteCtxt` is non
+// trivial
+/// Given some pending requests and a list of completed requests determine if
+/// we have another request the we can process
+fn get_next_benchmark_request(
+    pending: &mut [BenchmarkRequest],
+    completed: &[BenchmarkRequest],
+) -> Option<BenchmarkRequest> {
+    // We now know we are only looking at commits which are possibly waiting
+    // for their parent to be complete.
+    pending.sort_by(|a, b| {
+        a.commit_type
+            .rank()
+            .cmp(&b.commit_type.rank())
+            .then_with(|| a.pr().unwrap_or(&0).cmp(b.pr().unwrap_or(&0)))
+            .then_with(|| a.created_at.cmp(&b.created_at))
+            .then_with(|| a.tag().cmp(b.tag()))
+    });
+
+    let completed_set: HashSet<String> = completed.iter().map(|r| r.tag().to_string()).collect();
+
+    let cutoff = Utc::now() - chrono::Duration::days(30);
+
+    let pending_by_sha: HashMap<String, &BenchmarkRequest> =
+        pending.iter().map(|r| (r.tag().to_string(), r)).collect();
+
+    for req in pending.iter() {
```

Comment:
Interesting, we actually don't need to build an explicit topological ordering here and do the recursive parent lookup, because we can simply take a look if the parent is completed or not. Nevertheless, we will need to build the actual full queue somewhere, because we should show it on the status page, and we also need it for estimating how long it will take to benchmark a specific PR.

Reply:
We discussed this and I've implemented essentially what we have now with some minor modifications along with a test showing the ordering to be the same; https://github.com/rust-lang/rustc-perf/pull/2166/commits/6093f88dfcdd4219244f8d1b66d732257bd58de5